src/csharp/Microsoft.Spark/Sql/StorageLevel.cs

```diff
@@ -0,0 +1,238 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using Microsoft.Spark.Interop;
+using Microsoft.Spark.Interop.Ipc;
+
+namespace Microsoft.Spark.Sql
+{
+    /// <summary>
+    /// Flags for controlling the storage of an RDD. Each StorageLevel records whether to use
+    /// memory, whether to drop the RDD to disk if it falls out of memory, whether to keep the
+    /// data in memory in a JAVA-specific serialized format, and whether to replicate the RDD
+    /// partitions on multiple nodes. Also contains static properties for some commonly used
+    /// storage levels, MEMORY_ONLY.
+    /// </summary>
+    public sealed class StorageLevel : IJvmObjectReferenceProvider
+    {
+        private static readonly string s_storageLevelClassName =
+            "org.apache.spark.storage.StorageLevel";
+        private static StorageLevel s_none;
+        private static StorageLevel s_diskOnly;
+        private static StorageLevel s_diskOnly2;
+        private static StorageLevel s_memoryOnly;
+        private static StorageLevel s_memoryOnly2;
+        private static StorageLevel s_memoryOnlySer;
+        private static StorageLevel s_memoryOnlySer2;
+        private static StorageLevel s_memoryAndDisk;
+        private static StorageLevel s_memoryAndDisk2;
+        private static StorageLevel s_memoryAndDiskSer;
+        private static StorageLevel s_memoryAndDiskSer2;
+        private static StorageLevel s_offHeap;
+        private readonly JvmObjectReference _jvmObject;
+        private bool? _useDisk;
+        private bool? _useMemory;
+        private bool? _useOffHeap;
+        private bool? _deserialized;
+        private int? _replication;
```

Comment:
Are we caching these values?  If so, you should still set them in the constructor.

Reply:
I thought we wanted them to not be pre-populated in constructor, but set when called by user. Please correct me if I misunderstood @imback82

Reply:
We wanted to not pre-populate them by invoking the jvm.  The other constructor was doing this.  If the value is already passed in, it should be fine to cache them since you are planning to cache the jvm calls.

Reply:
Gotcha. Cool, will make the change. Thanks!

Reply:
My original intention was not to cache anything since this is not in a critical path to keep things simple. But either way is fine. :)